### DISCUSSION TITLE
Keep connection alive

### DISCUSSION THREAD

**Discussion by Fernando473** [deepgram/community#206] [General help] [answered]
How I can keep the connection alive with deepgram? 

I am using Django as backend and django-channels to connect with deepgram service.
I want to keep the connection with deepgram until close the socket with the client.

I have this code: 

```
from channels.generic.websocket import AsyncWebsocketConsumer
from deepgram import Deepgram
from typing import Dict


class TranscriptConsumer(AsyncWebsocketConsumer):
    dg_client = Deepgram("**********************************")
    audio_type = ''
    language = ''

    async def get_transcript(self, data: Dict) -> None:
        if 'channel' in data:
            transcript = data['channel']['alternatives'][0]['transcript']

            if transcript:
                await self.send(transcript)

    async def connect_to_deepgram(self):
        try:
            # Verify that the client and language are set
            if not self.dg_client:
                raise ValueError("Deepgram is not set.")
            if not self.language:
                raise ValueError("Language is not set.")

            # Establish connection with Deepgram
            self.socket = await self.dg_client.transcription.live(
                {'punctuate': True, 'interim_results': False, 'language': self.language})

            # Set timeout for socket
            self.socket.MESSAGE_TIMEOUT = 60.0

            # Register handlers
            self.socket.register_handler(
                self.socket.event.TRANSCRIPT_RECEIVED, self.get_transcript)
            self.socket.register_handler(
                self.socket.event.CLOSE, lambda c: print(f'Connection closed with code {c}.'))

        except ValueError as ve:
            raise ve

        except Exception as e:
            raise Exception(f'No se pudo abrir el socket: {e}')

    async def connect(self):        # Recieve protocols
        self.audio_type = self.scope['url_route']['kwargs']['audio_type']
        self.language = self.scope['url_route']['kwargs']['language']
        await self.connect_to_deepgram()
        await self.accept()

    async def disconnect(self, close_code):
        pass

    async def receive(self, bytes_data):
        self.socket.send(bytes_data)
```

I hope that you can help me

**Comment by nikolawhallon** [comment]
Deepgram has some timeouts if you don't send any audio for a while (10s) - however, if you don't have audio to send during that time, we have a new feature that allows you to keep the connection alive. Sending a text message with the following JSON:
`{"type":"KeepAlive"}`
will keep the connection alive for another ~10s, so sending this message once every ~< 10s should keep the connection alive even if there is no audio to stream during that time.

**Reply by nikolawhallon** [reply]
And here are docs for the feature: https://developers.deepgram.com/reference/streaming#stream-keepalive

**Comment by Fernando473** [comment]
You mean that I have to send some like that ?

` self.socket = await self.dg_client.transcription.live(
            {'punctuate': True, 'interim_results': False, 'language': self.language , "type":"KeepAlive" })`

**Reply by jjmaldonis** [reply, marked as answer]
Hey @Fernando473, instead of replacing `self.socket`, you want to send a message to that existing socket/connection.

So you can run `self.socket.send({"type": "KeepAlive"})` every 5-10 seconds to keep the connection open. When you're ready to close the connection, you can run `self.socket.send({"type": "CloseStream"})`.
